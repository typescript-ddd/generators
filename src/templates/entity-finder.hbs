import { Criteria, DomainEventBus, EntityNotFoundError } from "@typescript-ddd/core";
import { {{pascal entityName}}Repository } from "../../infrastructure/persistence";
import { {{pascal entityName}}, {{pascal entityName}}Id } from "../../domain";

/**
 * Represents a class to find {{plural (lower entityName)}}.
 */
export class {{pascal entityName}}Finder {
  /**
   * Initializes a new instance of {{noun entityName}} finder.
   * @param {{lbrace}}{{pascal entityName}}Repository{{rbrace}} repository - The {{natural entityName}} repository.
   * @param {DomainEventBus} eventBus - The domain event bus.
   */
  constructor(
    private readonly repository: {{pascal entityName}}Repository,
    private readonly eventBus: DomainEventBus
  ) {}

  /**
   * Finds {{noun entityName}} by a {{lower entityName}} id.
   * @param {{lbrace}}{{pascal entityName}}Id{{rbrace}} id - The {{natural entityName}} identity value.
   * @returns {{lbrace}}Promise<{{pascal entityName}}View>{{rbrace}} - A {{lower entityName}} view.
   */
  public async findById(id: {{pascal entityName}}Id): Promise<{{pascal entityName}}View> {
    const entity = await this.repository.findById(id);
    if (!entity) {
      throw new EntityNotFoundError("{{pascal entityName}}", id);
    }
    return {{pascal entityName}}View.create(...entity);
  }

  /**
   * Finds {{plural (lower entityName)}} by criteria.
   * @param {{lbrace}}Criteria{{rbrace}} criteria - The critieria value.
   * @returns {{lbrace}}Promise<{{pascal entityName}}View[]>{{rbrace}} - A collection of {{lower entityName}} views.
   */
  public async findAll(criteria: Criteria): Promise<{{pascal entityName}}View[]> {
    const entities = await this.repository.findAll(criteria);
    return entities.map((entity) => {{pascal entityName}}View.create(...entity));
  }
}
