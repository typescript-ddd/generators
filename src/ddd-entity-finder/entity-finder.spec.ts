import { EntityFinderTemplateData } from "./entity-finder.types";
import { generateEntityFinder } from "./entity-finder.generator";

describe("EntityCreator", () => {
  it("should render a created entity", () => {
    // Arrange
    const data: EntityFinderTemplateData = {
      entityName: "User",
    };

    // Act
    const result = generateEntityFinder(data);

    // Assert
    expect(result).toMatchSnapshot();
  });
});
