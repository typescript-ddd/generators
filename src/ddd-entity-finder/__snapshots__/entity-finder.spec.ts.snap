// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntityCreator should render a created entity 1`] = `
"import { Criteria, DomainEventBus, EntityNotFoundError } from "@typescript-ddd/core";
import { UserRepository } from "../../infrastructure/persistence";
import { User, UserId } from "../../domain";

/**
 * Represents a class to find users.
 */
export class UserFinder {
  /**
   * Initializes a new instance of a user finder.
   * @param {UserRepository} repository - The user repository.
   * @param {DomainEventBus} eventBus - The domain event bus.
   */
  constructor(
    private readonly repository: UserRepository,
    private readonly eventBus: DomainEventBus
  ) {}

  /**
   * Finds a user by a user id.
   * @param {UserId} id - The user identity value.
   * @returns {Promise<UserView>} - A user view.
   */
  public async findById(id: UserId): Promise<UserView> {
    const entity = await this.repository.findById(id);
    if (!entity) {
      throw new EntityNotFoundError("User", id);
    }
    return UserView.create(...entity);
  }

  /**
   * Finds users by criteria.
   * @param {Criteria} criteria - The critieria value.
   * @returns {Promise<UserView[]>} - A collection of user views.
   */
  public async findAll(criteria: Criteria): Promise<UserView[]> {
    const entities = await this.repository.findAll(criteria);
    return entities.map((entity) => UserView.create(...entity));
  }
}
"
`;
